ostd/src/mm/frame/linked_list.rs

```diff
@@ -0,0 +1,674 @@
+                abort();
+            }
+            self.list_id = id;
+            id
+        } else {
+            self.list_id
+        }
+    }
+}
+
+/// A cursor that can mutate the linked list links.
+///
+/// The cursor points to either a frame or the "ghost" non-element. It points
+/// to the "ghost" non-element when the cursor surpasses the back of the list.
+pub struct CursorMut<'a, M>
+where
+    Link<M>: AnyFrameMeta,
+{
+    list: &'a mut LinkedList<M>,
+    current: Option<NonNull<Link<M>>>,
+}
+
+impl<M> CursorMut<'_, M>
+where
+    Link<M>: AnyFrameMeta,
+{
+    /// Moves the cursor to the next frame towards the back.
+    ///
+    /// If the cursor is pointing to the "ghost" non-element then this will
+    /// move it to the first element of the [`LinkedList`]. If it is pointing
+    /// to the last element of the LinkedList then this will move it to the
+    /// "ghost" non-element.
+    pub fn move_next(&mut self) {
+        self.current = match self.current {
+            // SAFETY: The cursor is pointing to a valid element.
+            Some(current) => unsafe { *current.as_ref().next.get() },
+            None => self.list.front,
+        };
+    }
+
+    /// Moves the cursor to the previous frame towards the front.
+    ///
+    /// If the cursor is pointing to the "ghost" non-element then this will
+    /// move it to the last element of the [`LinkedList`]. If it is pointing
+    /// to the first element of the LinkedList then this will move it to the
+    /// "ghost" non-element.
+    pub fn move_prev(&mut self) {
+        self.current = match self.current {
+            // SAFETY: The cursor is pointing to a valid element.
+            Some(current) => unsafe { *current.as_ref().prev.get() },
+            None => self.list.back,
+        };
+    }
+
+    /// Gets the mutable reference to the current frame.
+    pub fn current(&mut self) -> Option<UniqueFrameRefMut<'_, Link<M>>> {
+        self.current.map(|current| {
+            let paddr = mapping::meta_to_frame::<PagingConsts>(current.as_ptr() as Vaddr);
+            // SAFETY: `&mut self` ensures we have exclusive access to the frame.
+            unsafe { UniqueFrameRefMut::from_raw(paddr) }
+        })
+    }
+
+    /// Takes the current pointing frame out of the linked list.
+    ///
+    /// If successful, the frame is returned and the cursor is moved to the
+    /// next frame. If the cursor is pointing to the back of the list then it
+    /// is moved to the "ghost" non-element.
+    pub fn take_current(&mut self) -> Option<UniqueFrame<Link<M>>> {
+        let frame = self.current.map(|current| {
+            let meta_ptr = current.as_ptr() as *mut MetaSlot;
+            let paddr = mapping::meta_to_frame::<PagingConsts>(meta_ptr as Vaddr);
+            // SAFETY: The frame was forgotten when inserted into the linked list.
+            unsafe { UniqueFrame::<Link<M>>::from_raw(paddr) }
+        })?;
+        // SAFETY: We own the links with `&mut self`.
```

Comment:
It might be better to insert a blank line between lines 287 and 288, because `frame` is referenced in several blocks of code below.